packages/settings/src/settings-feature-account-generate-vanity.tsx

```diff
@@ -0,0 +1,263 @@
+
+type VanityGeneratorState =
+  | { attempts: number; error: string | null; result: null; status: 'idle' | 'pending' | 'error' }
+  | { attempts: number; error: null; result: { address: string; secretKey: string }; status: 'success' }
+
+type VanityGeneratorAction =
+  | { type: 'start' }
+  | { payload: number; type: 'progress' }
+  | { payload: string; type: 'error' }
+  | { payload: { address: string; attempts: number; secretKey: string }; type: 'success' }
+  | { type: 'reset' }
+
+const initialVanityGeneratorState: VanityGeneratorState = { attempts: 0, error: null, result: null, status: 'idle' }
+
+function vanityGeneratorReducer(state: VanityGeneratorState, action: VanityGeneratorAction): VanityGeneratorState {
+  switch (action.type) {
+    case 'start':
+      return { attempts: 0, error: null, result: null, status: 'pending' }
+    case 'progress':
+      return { ...state, attempts: action.payload }
+    case 'success':
+      return {
+        attempts: action.payload.attempts,
+        error: null,
+        result: { address: action.payload.address, secretKey: action.payload.secretKey },
+        status: 'success',
+      }
+    case 'error':
+      return { attempts: 0, error: action.payload, result: null, status: 'error' }
+    case 'reset':
+      return initialVanityGeneratorState
+    default:
+      return state
+  }
+}
+
+function useVanityGenerator() {
+  const workerRef = useRef<Worker | null>(null)
+  const [state, dispatch] = useReducer(vanityGeneratorReducer, initialVanityGeneratorState)
+
+  useEffect(() => {
+    return () => {
+      workerRef.current?.terminate()
+      workerRef.current = null
+    }
+  }, [])
+
+  const start = async (input: VanityWalletFormFields) => {
+    const sanitizedPrefix = input.prefix?.trim().slice(0, 4) ?? ''
+    const sanitizedSuffix = input.suffix?.trim().slice(0, 4) ?? ''
+    const payload = {
+      caseSensitive: input.caseSensitive ?? true,
+      prefix: sanitizedPrefix,
+      suffix: sanitizedSuffix,
+    }
+
+    dispatch({ type: 'start' })
+    workerRef.current?.terminate()
+    const worker = new Worker(new URL('./workers/vanity-worker.ts', import.meta.url), { type: 'module' })
+    workerRef.current = worker
+
+    worker.onmessage = (event) => {
+      const { type, payload } = event.data ?? {}
+      if (type === 'progress' && typeof payload === 'number') {
+        dispatch({ payload, type: 'progress' })
+        return
+      }
+      if (type === 'found' && payload) {
+        dispatch({
+          payload: {
+            address: payload.address,
+            attempts: typeof payload.attempts === 'number' ? payload.attempts : 0,
+            secretKey: payload.secretKey,
+          },
+          type: 'success',
+        })
+        workerRef.current?.terminate()
+        workerRef.current = null
+        return
+      }
+      if (type === 'error') {
+        dispatch({
+          payload: typeof payload === 'string' ? payload : 'Failed to generate vanity address',
```

Comment:
Can you use type inference here? We should know the type of payload within this if statement